### PR TITLE
Replace python-jose with pyjwt

### PR DESCRIPTION
**Pull request by blutril**
## Purpose

Closing security gaps caused by python-jose.
https://www.first.org/cvss/calculator/3.1#CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:H/I:H/A:N
https://www.cve.org/CVERecord?id=CVE-2024-23342
https://www.cve.org/CVERecord?id=CVE-2024-33664
https://www.cve.org/CVERecord?id=CVE-2024-33663

## Does this introduce a breaking change?

When developers merge from main and run the server, azd up, or azd deploy, will this produce an error?
If you're not sure, try it out on an old environment.

```
[ ] Yes
[X] No
```

## Does this require changes to learn.microsoft.com docs?

This repository is referenced by [this tutorial](https://learn.microsoft.com/azure/developer/python/get-started-app-chat-template)
which includes deployment, settings and usage instructions. If text or screenshot need to change in the tutorial,
check the box below and notify the tutorial author. A Microsoft employee can do this for you if you're an external contributor.

```
[ ] Yes
[X ] No
```

## Type of change

```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/azure-search-openai-demo/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [X] The current tests all pass (`python -m pytest`).
- [x] I added tests that prove my fix is effective or that my feature works
- [x] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [x] I ran `python -m mypy` to check for type errors
- [X] I either used the pre-commit hooks or ran `ruff` and `black` manually on my code.

### CODE REVIEW

**Comment by pamelafox** [comment]
Thanks! I thought we'd investigated other package options, but don't remember whether jwt was considered. Asking @mattgotteiner to review as the original user of python-jose.

**Comment by pamelafox** [comment]
The mypy check failed, claims that "get_unverified_claims" doesn't exist in jwt. I just looked through their repository, and I don't see it either. So this seems to not be a viable replacement package for python-jose?

**Review by pamelafox — CHANGES_REQUESTED** [review]
See my comment re get_unverified_claims

**Review by mattgotteiner — APPROVED** [review]

**Review by pamelafox — APPROVED** [review]

**Comment by pamelafox** [comment]
@blutril Thanks so much! Merging.

**Comment by mattgotteiner** [comment]
Agreed - thanks @blutril  for resolving these security issues

**Comment by pamelafox** [comment]
Actually, I tested this out locally in my auth environment, and got this error:

```
  File "/workspaces/azure-search-openai-demo/.venv/lib/python3.11/site-packages/jwt/api_jwt.py", line 151, in decode_complete
    decoded = api_jws.decode_complete(
              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/azure-search-openai-demo/.venv/lib/python3.11/site-packages/jwt/api_jws.py", line 209, in decode_complete
    self._verify_signature(signing_input, header, signature, key, algorithms)
  File "/workspaces/azure-search-openai-demo/.venv/lib/python3.11/site-packages/jwt/api_jws.py", line 307, in _verify_signature
    prepared_key = alg_obj.prepare_key(key)
                   ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/workspaces/azure-search-openai-demo/.venv/lib/python3.11/site-packages/jwt/algorithms.py", line 341, in prepare_key
    raise TypeError("Expecting a PEM-formatted key.")
TypeError: Expecting a PEM-formatted key.
```

Have you tested this outside of the unit tests?

**Review by pamelafox — CHANGES_REQUESTED** [review]
This doesn't seem to work yet.

**Comment by TaylorN15** [comment]
> Actually, I tested this out locally in my auth environment, and got this error:
> 
> ```
>   File "/workspaces/azure-search-openai-demo/.venv/lib/python3.11/site-packages/jwt/api_jwt.py", line 151, in decode_complete
>     decoded = api_jws.decode_complete(
>               ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/azure-search-openai-demo/.venv/lib/python3.11/site-packages/jwt/api_jws.py", line 209, in decode_complete
>     self._verify_signature(signing_input, header, signature, key, algorithms)
>   File "/workspaces/azure-search-openai-demo/.venv/lib/python3.11/site-packages/jwt/api_jws.py", line 307, in _verify_signature
>     prepared_key = alg_obj.prepare_key(key)
>                    ^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/workspaces/azure-search-openai-demo/.venv/lib/python3.11/site-packages/jwt/algorithms.py", line 341, in prepare_key
>     raise TypeError("Expecting a PEM-formatted key.")
> TypeError: Expecting a PEM-formatted key.
> ```
> 
> Have you tested this outside of the unit tests?

I also tried using `pyjwt` to replace `python-jose` a little while back and encountered this same issue :(

**Comment by pamelafox** [comment]
@TaylorN15 Did you try their latest update? I just tested it, and it's working for me locally. I am testing in production next.

**Comment by pamelafox** [comment]
This also works for me in production. It would be great if a test could be added to verify the recently added code (i.e. a test that would have failed before your change to add RSA formatting).

**Comment by TaylorN15** [comment]
> @TaylorN15 Did you try their latest update? I just tested it, and it's working for me locally. I am testing in production next.

Just tested, works now :)

**Comment by pamelafox** [comment]
@blutril I tested locally and found a few more issues. I fixed them, modified existing test, added new test. Am now deploying it to prod but I suspect it'll be fine giving it's working locally. However, I couldn't push my fixes since you're working off your main branch, so could you please apply my changes from this commit?
https://github.com/pamelafox/azure-search-openai-demo/commit/3865c774c40afb90fa576fc487b24354d81040eb

**Review by pamelafox — CHANGES_REQUESTED** [review]
Please incorporate my commit-
https://github.com/pamelafox/azure-search-openai-demo/commit/3865c774c40afb90fa576fc487b24354d81040eb

**Review by pamelafox — APPROVED** [review]
This looks good, and is working for me locally and in production. Will ask @mattgotteiner to give it another look.
Given this change is security-related, we are putting it through more reviews than usual. Thanks for your patience.

File: app/backend/core/authentication.py [hunk 1]
```diff
@@ -304,51 +324,38 @@ async def validate_access_token(self, token: str):
                         jwks = await resp.json()
 
         if not jwks or "keys" not in jwks:
-            raise AuthError({"code": "invalid_keys", "description": "Unable to get keys to validate auth token."}, 401)
```
Comment (mattgotteiner) [inline app/backend/core/authentication.py:307]: note: we are switching away from json errors here

**Review by mattgotteiner — APPROVED** [review]

File: app/backend/core/authentication.py (hunk 1 above)
Comment (pamelafox) [reply app/backend/core/authentication.py:307]: I looked through the file and there seemed to be an inconsistent mix of JSON and non-JSON errors. So I debated whether to json.dumps() or just pass a string. We could json.dumps() if you think that's the better choice.

**Pull request merged**